src/Arcus.WebApi.Correlation/CorrelationMiddleware.cs

```diff
@@ -0,0 +1,130 @@
+            ILogger<CorrelationMiddleware> logger)
+        {
+            Guard.NotNull(next, nameof(next), "Requires a continuation delegate");
+            Guard.NotNull(options, nameof(options), "Requires a set of correlation options to manipulate how the correlation should happen");
+            Guard.NotNull(logger, nameof(logger), "Requires a logging implementation to trace diagnostic messages during the correlation");
+            Guard.For<ArgumentException>(() => options.Value is null, "Requires a set of correlation options to manipulate how the correlation should happen");
+
+            _next = next;
+            _options = options.Value;
+            _logger = logger;
+        }
+        
+        /// <summary>Request handling method.</summary>
+        /// <param name="httpContext">The <see cref="T:Microsoft.AspNetCore.Http.HttpContext" /> for the current request.</param>
+        /// <returns>A <see cref="T:System.Threading.Tasks.Task" /> that represents the execution of this middleware.</returns>
+        public async Task Invoke(HttpContext httpContext)
+        {
+            Guard.NotNull(httpContext, nameof(httpContext));
+            Guard.For<ArgumentException>(() => httpContext.Response is null, "Requires a 'Response'");
+            Guard.For<ArgumentException>(() => httpContext.Response.Headers is null, "Requires a 'Response' object with headers");
+
+            if (httpContext.Request.Headers.TryGetValue(_options.Transaction.HeaderName, out StringValues transactionIds))
+            {
+                if (!_options.Transaction.AllowInRequest)
+                {
+                    _logger.LogError("No correlation request header '{HeaderName}' for transaction ID was allowed in request", _options.Transaction.HeaderName);
+                    httpContext.Response.StatusCode = StatusCodes.Status400BadRequest;
+                    await httpContext.Response.WriteAsync($"No correlation transaction ID request header '{_options.Transaction.HeaderName}' was allowed in the request");
+
+                    return;
+                }
+
+                _logger.LogTrace("Correlation request header '{HeaderName}' found with transaction ID '{TransactionId}'", _options.Transaction.HeaderName, transactionIds);
+            }
+
+            string operationId = DetermineOperationId(httpContext);
+            string transactionId = DetermineTransactionId(httpContext, transactionIds);
+            var correlation = new CorrelationInfo(operationId, transactionId);
+            httpContext.Features.Set(correlation);
+
+            AddCorrelationResponseHeaders(httpContext, operationId, transactionId);
+
+            // TODO: on exception, should we enrich the exception.Data with the correlation info?
+            await _next(httpContext);
+        }
+
+        private static string DetermineOperationId(HttpContext httpContext)
+        {
+            // TODO: make ID generation configurable for the consumer.
```

Comment:
Let's add an issue for this and remove the todo?

Reply:
Done #102